compiler-plugin/src/main/java/io/ballerina/lib/data/csvdata/compiler/CsvDataTypeValidator.java

```diff
@@ -0,0 +1,649 @@
+
+        recordTypeSymbol.typeInclusions().forEach(symbol ->
+                validateNestedTypeSymbols(ctx, currentLocation, symbol, true));
+
+        recordTypeSymbol.fieldDescriptors().values().forEach(field -> validateNestedTypeSymbols(ctx,
+                currentLocation, field.typeDescriptor(), true));
+
+        Optional<TypeSymbol> restSymbol = recordTypeSymbol.restTypeDescriptor();
+        if (restSymbol.isPresent()) {
+            TypeSymbol restSym = restSymbol.get();
+            validateNestedTypeSymbols(ctx, currentLocation, restSym, true);
+        }
+    }
+
+    private void validateNestedTypeSymbols(SyntaxNodeAnalysisContext ctx,
+                                           Location location, TypeSymbol typeSymbol, boolean isField) {
+        TypeDescKind kind = typeSymbol.typeKind();
+        if (kind == TypeDescKind.TYPE_REFERENCE) {
+            kind = ((TypeReferenceTypeSymbol) typeSymbol).typeDescriptor().typeKind();
+        }
+
+        switch (kind) {
+            case ARRAY, OBJECT, RECORD, MAP, ERROR, FUNCTION, TUPLE, STREAM, FUTURE, TYPEDESC,
+                 TYPE_REFERENCE, XML, XML_ELEMENT, XML_PROCESSING_INSTRUCTION, XML_COMMENT,
+                 XML_TEXT, HANDLE, TABLE, NEVER, REGEXP  ->
+                 reportDiagnosticInfo(ctx, Optional.ofNullable(location),
+                         isField ? CsvDataDiagnosticCodes.UNSUPPORTED_FIELD_TYPE
+                        : CsvDataDiagnosticCodes.UNSUPPORTED_TUPLE_MEMBER_TYPE);
+        }
+    }
+
+    public static TypeSymbol getRawType(TypeSymbol typeDescriptor) {
+        if (typeDescriptor.typeKind() == TypeDescKind.INTERSECTION) {
+            return getRawType(((IntersectionTypeSymbol) typeDescriptor).effectiveTypeDescriptor());
+        }
+        if (typeDescriptor.typeKind() == TypeDescKind.TYPE_REFERENCE) {
+            TypeReferenceTypeSymbol typeRef = (TypeReferenceTypeSymbol) typeDescriptor;
+            if (typeRef.typeDescriptor().typeKind() == TypeDescKind.INTERSECTION) {
+                return getRawType(((IntersectionTypeSymbol) typeRef.typeDescriptor()).effectiveTypeDescriptor());
+            }
+            TypeSymbol rawType = typeRef.typeDescriptor();
+            if (rawType.typeKind() == TypeDescKind.TYPE_REFERENCE) {
+                return getRawType(rawType);
+            }
+            return rawType;
+        }
+        return typeDescriptor;
+    }
+
+    private void reportDiagnosticInfo(SyntaxNodeAnalysisContext ctx, Optional<Location> location,
+                                      CsvDataDiagnosticCodes diagnosticsCodes) {
+        Location pos = location.orElseGet(() -> currentLocation);
+        DiagnosticInfo diagnosticInfo = new DiagnosticInfo(diagnosticsCodes.getCode(),
+                diagnosticsCodes.getMessage(), diagnosticsCodes.getSeverity());
+        if (pos == null || (allDiagnosticInfo.containsKey(pos) && allDiagnosticInfo.get(pos).equals(diagnosticInfo))) {
```

Comment:
Given diagnosticInfo is always non-null I think you can avoid doing a lookup on the map twice here
```suggestion
        if (pos == null || Objectes.equals(allDiagnosticInfo.get(pos), diagnosticInfo)) {
```